security: Add support for OCSP

bdarnell opened the pull request:
This adds support for certificate revocation for organizations who run
an OCSP server for their CA.

Closes #29641

Release note (security update): Certificate revocation is now
supported via OSCP and the setting `security.ocsp.mode`.

Still a WIP: includes the plumbing and the implementation, but I'm still working on tests.

bdarnell committed:
security: Add support for OCSP

This adds support for certificate revocation for organizations who run
an OCSP server for their CA.

This is an MVP implementation of revocation checking; it does not
implement caching and incurs a call to the OCSP server on every
connection attempt.

Closes #29641

Release note (security update): Certificate revocation is now
supported via OSCP and the setting `security.ocsp.mode`. This makes a
call to the OCSP server on every connection attempt and may wait for
up to the setting `security.ocsp.timeout`.

knz reviewed (commented):
Thanks for this.
I notice that this does not implement caching as outlined in the RFC so you may want to call out the performance impact in the release note.

> Reviewed 18 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFKA-NP55ab4cUrpeN-:b-it27x7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)

___
*[pkg/security/ocsp.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9gejAE7jRiFb_C8k:-MFK9gek2SvOL4AFz0v-:bs5an5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L126)):*
> ```Go
> 
> 	if httpResp.StatusCode != http.StatusOK {
> 		return false, errors.Newf("OCSP server returned status code %v", httpResp.StatusCode)
> ```

can you add an `errors.Safe` around that status code
___
*[pkg/security/ocsp.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9kT014bA-WQ72itO:-MFK9kT014bA-WQ72itP:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L129)):*
> ```Go
> 	}
> 	if ct := httpResp.Header.Get("Content-Type"); ct != "application/ocsp-response" {
> 		return false, errors.Newf("OCSP server returned unexpected content-type %q", ct)
> ```

ditto
___
*[pkg/security/ocsp.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9muP0LRHU2l_5dBI:-MFK9muP0LRHU2l_5dBJ:bg3thyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L150)):*
> ```Go
> 		return false, nil
> 	default:
> 		return false, errors.Newf("OCSP returned status %v", ocspResp.Status)
> ```

Status "unknown" is part of the spec should not be an error IMHO. 
unknown + lax = good
unknown + strict = revoked 


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
Added a test. Will address other comments tomorrow. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFNr-Ph6MHhV3A7pLPC:blnwvfl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, @knz, and @tbg)

___
*[pkg/security/ocsp.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9muP0LRHU2l_5dBI:-MFNr5Dt5FdAKbsiW9cd:b8quam2) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Status "unknown" is part of the spec should not be an error IMHO. 
unknown + lax = good
unknown + strict = revoked 
</blockquote></details>

That's exactly what happens: we return an error here and then our caller checks the strict flag to decide whether the cert is accepted or not. 


<!-- Sent from Reviewable.io -->

bdarnell commented:
Rebased and CI is green; ready for final review. I've added errors.Safe in a few places. I don't think the release note is the place to talk about the performance limitations of this implementation but I'll make sure it's covered when the feature is documented. 

The test isn't the prettiest thing but every time I tried to factor out some repetitive bits I ran into my limited knowledge of tcl and I couldn't get things working the way I wanted to. 

> caching as outlined in the RFC

I didn't see much discussion of caching in RFC 6960. Are you referring to the use of HTTP GET in some cases, or is there something else I missed?

knz reviewed (commented):
Thanks, the testing looks fine.

> I don't think the release note is the place to talk about the performance limitations of this implementation

The release note is the place to talk about the performance impact of enabling OCSP **and** also the place to communicate high-level insights to the doc writer.
As-is this implementation is incurring up to N seconds penalty on every new connection from a client. The operator needs to be aware of this fact.

> I didn't see much discussion of caching in RFC 6960. [...] is there something else I missed?

Yes you more or less completely ignored what was discussed in #50602.

I think at the very least your commit message  should include the following sentences:

- in the release note: "This implementation is experimental and may evolve in the next version of CockroachDB."
- in the commit message: "This is a MVP implementation of the concepts outlined in the revocation RFC. This does not implement response caching and thus incurs a mandatory OCSP roundtrip upon every TLS handshake."

> Reviewed 15 of 15 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFUb2UeDMbLmpzO15nV:b-oofl9e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)

___
*[pkg/cli/interactive_tests/test_secure_ocsp.tcl, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFU_fv5-wjdyQ9VPw2J:-MFU_fv5-wjdyQ9VPw2K:b-ntp546) ([raw file](https://github.com/cockroachdb/cockroach/blob/5842f0c82756ecf71d0e8165dacc7d173621eaa3/pkg/cli/interactive_tests/test_secure_ocsp.tcl#L61)):*
> ```Tcl
> create_user_cert $argv $certs_dir gallant
> 
> report "REVOKING CERTIFICATE"
> ```

nit: we have `start_test` / `end_test` directives
___
*[pkg/cli/interactive_tests/test_secure_ocsp.tcl, line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFU_dG8E1gfUugUUy-a:-MFU_dG91jsxeerSgRNw:b-ahxfpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/5842f0c82756ecf71d0e8165dacc7d173621eaa3/pkg/cli/interactive_tests/test_secure_ocsp.tcl#L110)):*
> ```Tcl
> system "$argv sql --certs-dir=$certs_dir --user=goofus -e 'select 1'"
> 
> # One more time, strict mode with stopped server (so both fail
> ```

nit: missing parenthesis
___
*[pkg/security/ocsp.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9muP0LRHU2l_5dBI:-MFU_wBC3d8P0HxEguzp:b-59he2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L150)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

That's exactly what happens: we return an error here and then our caller checks the strict flag to decide whether the cert is accepted or not. 
</blockquote></details>

Your current implementation does not very well distinguish between "off" and "lax" in case the server is misconfigured (or, sadly, if the connection is being attacked via MITM). Is that intentional? I would have expected a hard error (and OCSP failure) if the setting is "lax" but the OCSP response is badly signed or the server cannot be contacted.
___
*[pkg/security/tls_settings.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFUap6X1FzCYxEVrmw6:-MFUap6X1FzCYxEVrmw7:b-rhd7kb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5842f0c82756ecf71d0e8165dacc7d173621eaa3/pkg/security/tls_settings.go#L43)):*
> ```Go
> 
> var ocspTimeout = settings.RegisterPublicNonNegativeDurationSetting("security.ocsp.timeout",
> 	"timeout before considering the OCSP server unreachable", time.Second)
> ```

A second is very short. 5 seconds maybe?
Also this needs to be called out in the release note.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> > I didn't see much discussion of caching in RFC 6960. [...] is there something else I missed?
> 
> Yes you more or less completely ignored what was discussed in #50602.

Oh, our RFC, not the IETF's. Got it. 

> in the release note: "This implementation is experimental and may evolve in the next version of CockroachDB."

If we call it experimental don't the setting names need to change? I don't think it's experimental in that way - adding caching and other tweaks should be pretty transparent.

Updated the release note. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFVHzkF8jH8SLl3DUAx:bty0jkd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @knz, and @tbg)

___
*[pkg/cli/interactive_tests/test_secure_ocsp.tcl, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFU_fv5-wjdyQ9VPw2J:-MFVIkPt2mOGA0reo8Ar:b-82a55q) ([raw file](https://github.com/cockroachdb/cockroach/blob/5842f0c82756ecf71d0e8165dacc7d173621eaa3/pkg/cli/interactive_tests/test_secure_ocsp.tcl#L61)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: we have `start_test` / `end_test` directives
</blockquote></details>

I think `report` is the right thing to use on this line, but changed all the `report "TESTING..."` lines to use start_test and end_test. 
___
*[pkg/cli/interactive_tests/test_secure_ocsp.tcl, line 110 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFU_dG8E1gfUugUUy-a:-MFVIjir5ijH5jhkT_CJ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5842f0c82756ecf71d0e8165dacc7d173621eaa3/pkg/cli/interactive_tests/test_secure_ocsp.tcl#L110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: missing parenthesis
</blockquote></details>

Done
___
*[pkg/security/ocsp.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9gejAE7jRiFb_C8k:-MFVIY0p4ePvhxuNHW7c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L126)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

can you add an `errors.Safe` around that status code
</blockquote></details>

Done.
___
*[pkg/security/ocsp.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9kT014bA-WQ72itO:-MFVIYT02aLXG3uhINb7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L129)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/security/ocsp.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9muP0LRHU2l_5dBI:-MFVJxMx-QIeksIqSULS:bi73re4) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L150)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Your current implementation does not very well distinguish between "off" and "lax" in case the server is misconfigured (or, sadly, if the connection is being attacked via MITM). Is that intentional? I would have expected a hard error (and OCSP failure) if the setting is "lax" but the OCSP response is badly signed or the server cannot be contacted.
</blockquote></details>

Yes, this is intentional - the point of lax mode is that a DoS (or MITM) attack on your OCSP server shouldn't become a DoS against everything that uses it. Lax mode is "block certs that are known to be bad"; strict mode is "allow only certs that are known to be good"; any sort of error prevents us from knowing anything. 
___
*[pkg/security/tls_settings.go, line 43 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFUap6X1FzCYxEVrmw6:-MFVIg6j3eb3Ibl7hqvs:b-qspjk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5842f0c82756ecf71d0e8165dacc7d173621eaa3/pkg/security/tls_settings.go#L43)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

A second is very short. 5 seconds maybe?
Also this needs to be called out in the release note.
</blockquote></details>

Changed to 3 seconds to match `base.NetworkTimeout` (but I can't use the constant because of import cycles, grumble)


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> If we call it experimental don't the setting names need to change?

We changed it last year sometime - the promotion from experimental to non-experimental should not force users to change their deployments. A callout in the description of the setting and/or the release note is sufficient.

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFVNHU518Jt6bP6Mto9:b-oncybz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)

___
*[pkg/security/ocsp.go, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-MFK9muP0LRHU2l_5dBI:-MFVN7Nr7dT8evgRrhKx:b-rs8zjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/228bf4a654cf29dc0ad7579babed1a3b6f4821a1/pkg/security/ocsp.go#L150)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, this is intentional - the point of lax mode is that a DoS (or MITM) attack on your OCSP server shouldn't become a DoS against everything that uses it. Lax mode is "block certs that are known to be bad"; strict mode is "allow only certs that are known to be good"; any sort of error prevents us from knowing anything. 
</blockquote></details>

thanks. I believe this should be explained in the commit message too (probably in the release note).


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
>  I don't think it's experimental in that way - adding caching and other tweaks should be pretty transparent.

We don't know yet whether this implementation is adequate - in particular whether we want a single OCSP setting for all the TLS checkers, or separate settings. I'd say the UX may not change in a backward-compatible but the look & feel and the expectations of users may change in the future. Hence my suggestion for a healthy dose of disclaimers.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFVNe4ECYN7d1eN93SL:bo9636y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> may not change in a backward-compatible 

in a backward-incompatible manner

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53218#-:-MFVNx9DDpUHilHvMBPt:b-z69cmb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl, @bdarnell, and @tbg)



<!-- Sent from Reviewable.io -->

bdarnell commented:
bors r=knz
